framework/src/transfers/MultiAppProjectionTransfer.C

```diff
@@ -220,44 +223,96 @@ MultiAppProjectionTransfer::assembleL2To(EquationSystems & es, const std::string
   {
-    if (!_multi_app->hasLocalApp(i))
+    if (!_multi_app->hasLocalApp(i_global))
       continue;
 
-    MPI_Comm swapped = Moose::swapLibMeshComm(_multi_app->comm());
-
-    FEProblem & from_problem = *_multi_app->appProblem(i);
+    // Get the bounding box.
+    FEProblem & from_problem = *_multi_app->appProblem(i_global);
     EquationSystems & from_es = from_problem.es();
     MeshBase & from_mesh = from_es.get_mesh();
-    MeshTools::BoundingBox * app_box = new MeshTools::BoundingBox(MeshTools::processor_bounding_box(from_mesh, from_mesh.processor_id()));
-    from_bbs[i] = app_box;
+    MeshTools::BoundingBox app_box = MeshTools::BoundingBox(MeshTools::processor_bounding_box(from_mesh, from_mesh.processor_id()));
+
+    // Cast the bounding box into a pair of points to simplify MPI
+    // communication.  Translate the bounding box to the app's position.
+    bb_points[i_local] = static_cast<std::pair<Point, Point> >(app_box);
+    bb_points[i_local].first = bb_points[i_local].first + _multi_app->position(i_global);
+    bb_points[i_local].second = bb_points[i_local].second + _multi_app->position(i_global);
+    local_bboxes[i_local] = static_cast<MeshTools::BoundingBox>(bb_points[i_local]);
 
+    // Get a serialized copy of the subapp's solution vector.
     MooseVariable & from_var = from_problem.getVariable(0, _from_var_name);
     System & from_sys = from_var.sys().system();
     unsigned int from_var_num = from_sys.variable_number(from_var.name());
 
-    NumericVector<Number> * serialized_from_solution = NumericVector<Number>::build(from_sys.comm()).release();
-    serialized_from_solution->init(from_sys.n_dofs(), false, SERIAL);
-    // Need to pull down a full copy of this vector on every processor so we can get values in parallel
-    from_sys.solution->localize(*serialized_from_solution);
-    from_slns[i] = serialized_from_solution;
+    serialized_from_solutions[i_local] = NumericVector<Number>::build(from_sys.comm()).release();
+    serialized_from_solutions[i_local]->init(from_sys.n_dofs(), false, SERIAL);
+    from_sys.solution->localize(*serialized_from_solutions[i_local]);
 
-    MeshFunction * from_func = new MeshFunction(from_es, *serialized_from_solution, from_sys.get_dof_map(), from_var_num);
+    // Get the subapp's mesh function.
+    MeshFunction * from_func = new MeshFunction(from_es, *serialized_from_solutions[i_local], from_sys.get_dof_map(), from_var_num);
     from_func->init(Trees::ELEMENTS);
     from_func->enable_out_of_mesh_mode(OutOfMeshValue);
-    from_fns[i] = from_func;
+    local_meshfuns[i_local] = from_func;
 
-    Moose::swapLibMeshComm(swapped);
+    i_local++;
   }
 
-
+  /********************
+  Next, serialize the bounding boxes, and keep track of how many apps (i.e. how
+  many bounding boxes) each processor has.
+  ********************/
+  std::vector<unsigned int> apps_per_proc(1, n_local_apps);
+
+  _communicator.allgather(bb_points);
+  _communicator.allgather(apps_per_proc, true);
```

Comment:
This one here is the one you should change.  It's doing a lot more work than is necessary for sending one value.